### PR TITLE
Add Assert feature

### PR DESCRIPTION
**Pull request by illegitimate-egg**
Based off of the GNU assert macros, if a statement/boolean is false then the assert will stop the program and print out debug information.

image

### CODE REVIEW

**Review by williamistGitHub — COMMENTED** [review]
Cool idea, but I don't think a separate log level just for asserts is warranted, DEBUG would work just fine. Hardcoding assert messages to always exit is a tad odd as well, but that's a non issue if you just use DEBUG (which you probably should be doing anyway).

I'm also unsure if a logging library is a good place to have an assertion macro, couldn't you just #define this in your code? As far as I can tell the only reason to put it here would be to have the separate log level, which as I previously stated I don't like.

If you have a real reason to include this here instead of in your code, feel free to let me know, otherwise this should be closed.

**Comment by illegitimate-egg** [comment]
Maybe this would be more viable as a seperate file? The assert exits because that's what the traditional beahviour defined in POSIX says that it should do. I think a logging library is the ideal place for something like this, because the only other place for it is in a standard library implementation, and formatting it like this would make it POSIX incompliant.

The additional log-level is addmittedly unneccessary and can be removed. I can add the exit call to the macro instead.

**Comment by williamistGitHub** [comment]
This is short enough I don't think it should be a part of this library. It's easy enough to copy-paste into your own code.

You're free to add an example showing that you can do this if you'd like, I'd accept that.

**Comment by illegitimate-egg** [comment]
Being short enough to paste into your own code goes for almost all of the standard library functions.

Personally I would get mileage out of this being part of CLog. Anyone can write a logging library, but what makes a library good is having the features you need when you need them.

**Comment by williamistGitHub** [comment]
Fair. Feel free to make a `clog/extra.h` with this macro in it, I have a feeling other useful stuff that doesn't quite fit the library's goal could go there as well in the future.

File: CMakeLists.txt
Comment (williamistGitHub) [inline CMakeLists.txt:1]: This is just unnecessary whitespace changes. What's the point of all this?

File: examples/CMakeLists.txt [hunk 1]
```diff
@@ -1,2 +1,4 @@
 add_executable(basic basic/basic.c)
-target_link_libraries(basic clog)
\ No newline at end of file
+add_executable(assert extra/assert.c)
+target_link_libraries(basic clog)
```
Comment (williamistGitHub) [inline examples/CMakeLists.txt:2]: Should probably move the `target_link_libraries` for `basic` next to it's executable creation, that way the two examples are fully separated.

File: examples/extra/assert.c [hunk 2]
```diff
@@ -0,0 +1,12 @@
+/*
+* CLog Assert Example
```
Comment (williamistGitHub) [inline examples/extra/assert.c:2]: Should add the CC0 disclaimer here as well, to be consistent with the basic example.

File: examples/extra/assert.c
Comment (williamistGitHub) [inline examples/extra/assert.c:1]: This file is mostly good, I mainly have an issue with the comments being C++ style `//` instead of C89 style `/**/`.

Not a big issue here since this is an example and not part of the library itself but I feel it'd be good to be consistent.

The other comment issue is the use of first person language (e.g. "I checked"), would prefer the use of more generic language like "Passing assertion example:" or something like that.

File: include/clog/extra.h [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+#define CLOG_EXTRA_H
+
+#include <stdlib.h>
+#include "clog.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+void __clog_assert(const char *msg, const char *file, int line) {
```
Comment (williamistGitHub) [inline include/clog/extra.h]: Defining a function in a header file is a very bad idea, especially if it isn't `inline`. Can you not include this in the macro itself?

File: include/clog/extra.h [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    clog_log(CLOG_LEVEL_ERROR, "%s at %s, line %d", msg, file, line);
+    exit(1); // POSIX says that assert should exit
```
Comment (williamistGitHub) [inline include/clog/extra.h]: As this header file *is* a part of the library, this comment will break C89 support and so must be changed before I can accept this. As I mentioned earlier though, I would like this stuff be moved into the macro anyway, which would remove the need for this function at all.

File: include/clog/extra.h [hunk 5]
```diff
@@ -0,0 +1,26 @@
+#ifndef CLOG_EXTRA_H
+#define CLOG_EXTRA_H
+
+#include <stdlib.h>
+#include "clog.h"
```
Comment (williamistGitHub) [inline include/clog/extra.h]: This should be `clog/clog.h` for consistency.

File: include/clog/extra.h [hunk 6]
```diff
@@ -0,0 +1,26 @@
+#ifndef CLOG_EXTRA_H
```
Comment (williamistGitHub) [inline include/clog/extra.h:24]: There should be the license header here, can just copy from `clog.h`.

**Review by williamistGitHub — CHANGES_REQUESTED** [review]
A good start, but needs improvement. Especially in `extra.h`.

File: CMakeLists.txt
Comment (illegitimate-egg) [reply CMakeLists.txt:1]: Sorry about that, my formatter was playing up, won't happen again.

File: examples/CMakeLists.txt (hunk 1 above)
Comment (illegitimate-egg) [reply examples/CMakeLists.txt:2]: I internalized it without segregation, first executables are assigned source files, then they are linked. They are processed one step for each at a time, rather than 1 at a time. You're right that they should be separate.

File: examples/extra/assert.c (hunk 2 above)
Comment (illegitimate-egg) [reply examples/extra/assert.c:2]: I'm not sure, will license.md not suffice?

File: examples/extra/assert.c
Comment (illegitimate-egg) [reply examples/extra/assert.c:1]: I wasn't aware of the difference in comments, will remember for the future

File: include/clog/extra.h (hunk 3 above)
Comment (illegitimate-egg) [reply include/clog/extra.h]: Inclusion within the macro isn't possible, it's an if statement, if statements can't execute multiple functions. This also can't be inlined*. The only option is to make a separate source file.

\* I don't know why it didn't work

File: include/clog/extra.h (hunk 4 above)
Comment (illegitimate-egg) [reply include/clog/extra.h]: Yeah that's on me

File: include/clog/extra.h (hunk 5 above)
Comment (illegitimate-egg) [reply include/clog/extra.h]: Alright

File: include/clog/extra.h (hunk 6 above)
Comment (illegitimate-egg) [reply include/clog/extra.h:24]: Done

File: CMakeLists.txt
Comment (williamistGitHub) [reply CMakeLists.txt:1]: Just revert the change and it'll be fine.

File: examples/CMakeLists.txt (hunk 1 above)
Comment (williamistGitHub) [reply examples/CMakeLists.txt:2]: I'm not hearing any good reason to do this, and I feel it'd be better to keep the different examples separate in the case that one needs to be modified it can all be done in one place.

File: examples/extra/assert.c (hunk 2 above)
Comment (williamistGitHub) [reply examples/extra/assert.c:2]: It's easy enough to do, just a one line change. If you want me to do it instead of you doing it I can.

File: examples/extra/assert.c
Comment (williamistGitHub) [reply examples/extra/assert.c:1]: No problem.

File: include/clog/extra.h (hunk 3 above)
Comment (williamistGitHub) [reply include/clog/extra.h]: Inclusion within the macro is most definitely possible:

```c
#define clog_assert(EX) (void)((EX) || ((clog_log(CLOG_LEVEL_DEBUG, "Assertion failed: %s at %s:%d", #EX, __FILE__, __LINE__)), exit(1), 0))
```

Comment (illegitimate-egg) [reply include/clog/extra.h]: Ah, I tried using a semicolon instead of a comma, I don't understand macros very well.

Comment (williamistGitHub) [reply include/clog/extra.h]: All good.

**Comment by illegitimate-egg** [comment]
I think it should meet your checks now

File: include/clog/extra.h [hunk 7]
```diff
@@ -1,24 +1,42 @@
...
-    exit(1); // POSIX says that assert should exit
-}
-
 /**
  * @brief A standin for the regular C assert macro
  *
  * @param EX The condition to check
  */
-#define clog_assert(EX) (void)((EX) || ((__clog_assert(#EX, __FILE__, __LINE__)), 0))
+#define clog_assert(EX) (void)((EX) || (clog_log(CLOG_LEVEL_ERROR, "%s at %s, line %d", #EX, __FILE__, __LINE__), exit(1), 0)) /* POSIX says that assert should exit */
```
Comment (williamistGitHub) [inline include/clog/extra.h]: Kind of a nitpick but the message could be better here. If nothing else, prefixing with `Assertion failed:` would be nice. Other than that, this file looks good.

**Review by williamistGitHub — COMMENTED** [review]
Definitely better.

**Comment by illegitimate-egg** [comment]
Should be all how you wanted it, I fixed up CMakeLists and added your indicator

**Review by williamistGitHub — APPROVED** [review]

**Comment by williamistGitHub** [comment]
Looks all good now, thanks :)

**Comment by illegitimate-egg** [comment]
Brilliant, when V2 is completed, this message can be overhauled to be a custom message rather than have `[Error] Assert Failed:` at the start of everything. Just [Assert Failed] will be nice.
